OpenCL: Implement CUDA callbacks using OCL CBs and user event hacks

pjaaskel opened the pull request:
hipStreamSemantics tests behavior that CUDA callbacks are asynchronously executed in the background whereas OpenCL might or might not execute them in the background. It could execute the callback in the user thread in case the command it is calling back from was finished before the CB was set.

Hack around this by creating a barrier which listens to a user event which is triggered only after setting the CB, forcing parallel execution of the callback. That is, in case the OpenCL platform supports parallel execution of callbacks the first place. I don't believe that's required to be the case, so this is down to implementation still.

Might Fix #421 and #418.


With this my local test envs now pass the expected tests once again.

pvelesko commented:
What about taking this a step further and lifting the L0 callback implementation to CHIP level and then just using that generic implementation for both L0 and OCL? 

That way we don't need the OpenCL implementation to support async callback execution and we have less backend divergence.

pvelesko commented:
Failing for all platforms: 
```
dgpu_opencl_make_check_result.txt: FAIL
	708 - Unit_hipStreamAddCallback_WithDefaultStream (Failed)
	709 - Unit_hipStreamAddCallback_WithCreatedStream (Failed)
```

Also, if there are bugs that this PR might fix please remove them from exclusion list so that they get run through CI.

pvelesko commented:
rebased

pjaaskel commented:
> What about taking this a step further and lifting the L0 callback implementation to CHIP level and then just using that generic implementation for both L0 and OCL?
> 
> That way we don't need the OpenCL implementation to support async callback execution and we have less backend divergence.
> 
> Not as a part of this PR but as a future follow-up

Yes, that crossed my mind, but then I started to think that it might be most efficient to push execution (decisions) down to driver level as much as possible so they can optimize the parallelism and synchronizations as they see best. Here we might need to resort to our own host thread anyhow because the OpenCL callbacks are so weakly defined. Let's see if this works out or not.

> 	708 - Unit_hipStreamAddCallback_WithDefaultStream (Failed)
>	709 - Unit_hipStreamAddCallback_WithCreatedStream (Failed)

These pass for my OpenCL envs (Intel iGPU, Intel CPU, PoCL-CPU and PoCL-Level0-iGPU) for me. What is the error output?

> Also, if there are bugs that this PR might fix please remove them from exclusion list so that they get run through CI.

I do not know any immediately, but we should anyhow start going through the exclusion lists and fix things systematically when stabilizing the 1.0. Do you have hunch for some tests which could be fixed by this?

pvelesko commented:
> Yes, that crossed my mind, but then I started to think that it might be most efficient to push execution (decisions) down to driver level as much as possible so they can optimize the parallelism and synchronizations as they see best. Here we might need to resort to our own host thread anyhow because the OpenCL callbacks are so weakly defined. Let's see if this works out or not.

Well if OpenCL doesn't guarantee async execution we kind of are bound by the HIP API to use a callback executor thread, no? 

> I do not know any immediately, but we should anyhow start going through the exclusion lists and fix things systematically when stabilizing the 1.0. Do you have hunch for some tests which could be fixed by this?

I have a workflow that will run for days testing each failure in the exclusion list for resolution.

pjaaskel commented:
> Well if OpenCL doesn't guarantee async execution we kind of are bound by the HIP API to use a callback executor thread, no?

Not sure of this, since I didn't find a mention in the CUDA PM that guarantees it either. There are other similar spots in terms of concurrency "assumptions" we discussed some 6 months ago that might not be true when running on various OpenCL implementations. I'm not sure if e.g. OpenCL mandates command queue async execution at all. We've created an implementation that doesn't need multithreading (or even fibers) for very simple microcontroller deployments (PoCL-basic), and it should be workable (the current implementation has a couple of issues though).

pjaaskel commented:
hipStreamAddCallback seems fishy, it lacks volatile in a global it uses to signal completion.

Pls try if this helps (it still works here):

```C++
diff --git a/tests/catch/unit/stream/hipStreamAddCallback.cc b/tests/catch/unit/stream/hipStreamAddCallback.cc
index 279f372d..b1b97099 100644
--- a/tests/catch/unit/stream/hipStreamAddCallback.cc
+++ b/tests/catch/unit/stream/hipStreamAddCallback.cc
@@ -39,7 +39,7 @@ Testcase Scenarios :
 namespace hipStreaAddCallbackTest {
 size_t NSize = 4 * 1024 * 1024;
 float *A_h, *C_h;
-bool gcbDone = false;
+volatile bool gcbDone = false;
 bool gPassed = true;
 void *ptr0xff = reinterpret_cast<void *>(0xffffffff);
 void *gusrptr;
```

pvelesko commented:
@pjaaskel  I ran the failed tests manually and they passed. On the CI pass they failed via a correctness issue.

pvelesko commented:
Failed again in the CI env:

```
 78/806 Test #708: Unit_hipStreamAddCallback_WithDefaultStream ...............................***Failed    0.70 sec
Filters: Unit_hipStreamAddCallback_WithDefaultStream
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
hipStreamAddCallback is a Catch v2.13.4 host application.
Run with -? for options
-------------------------------------------------------------------------------
Unit_hipStreamAddCallback_WithDefaultStream
-------------------------------------------------------------------------------
/home/pvelesko/actions-runner/_work/chip-spv/chip-spv/HIP/tests/catch/unit/stream/hipStreamAddCallback.cc:230
...............................................................................
/home/pvelesko/actions-runner/_work/chip-spv/chip-spv/HIP/tests/catch/unit/stream/hipStreamAddCallback.cc:233: FAILED:
  REQUIRE( TestPassed )
with expansion:
  false
with message:
  Data mismatch :58364
===============================================================================
test cases: 1 | 1 failed
assertions: 4 | 3 passed | 1 failed
        Start 650: Unit_hipMemsetFunctional_ZeroValue_3D
 79/806 Test #709: Unit_hipStreamAddCallback_WithCreatedStream ...............................***Failed    0.75 sec
Filters: Unit_hipStreamAddCallback_WithCreatedStream
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
hipStreamAddCallback is a Catch v2.13.4 host application.
Run with -? for options
-------------------------------------------------------------------------------
Unit_hipStreamAddCallback_WithCreatedStream
-------------------------------------------------------------------------------
/home/pvelesko/actions-runner/_work/chip-spv/chip-spv/HIP/tests/catch/unit/stream/hipStreamAddCallback.cc:239
...............................................................................
/home/pvelesko/actions-runner/_work/chip-spv/chip-spv/HIP/tests/catch/unit/stream/hipStreamAddCallback.cc:242: FAILED:
  REQUIRE( TestPassed )
with expansion:
  false
with message:
  Data mismatch :0
```

I'll try your patch but it's very strange that I can't reproduce it outside of CI

pjaaskel commented:
Yes, strange. I'd expect the same binary to fail in this case the same way as the volatile should be optimized by the compiler if something.

BTW, this is the description related to callbacks in CUDA in the concurrency Chapter:  https://docs.nvidia.com/cuda/archive/10.0/cuda-c-programming-guide/index.html#stream-callbacks

It doesn't mention the CBs should run in parallel. In the current implementation they usually should do, or they'll run in a blocking fashion when the holding user event is set CL_COMPLETE. I'd guess all implementations than PoCL-basic will execute it in parallel, as expected by the hipStreamSemantics tests (which might tests semantics that is not guaranteed to be there, if I read it right).

pjaaskel commented:
Is it happening only with dGPU?

pvelesko commented:
> Is it happening only with dGPU?

yes dGPU OpenCL. You can checkout the logs yourself.

pvelesko commented:
> The log is not very helpful. But I'll test with a dGPU when I get to one.

Would it be helpful to get access to the CI machine to test this?

pjaaskel commented:
OK, I reproduced it successfully in your box and fixed the problem (I think).

While testing, I noticed hipStreamSemantics is falsely detected as passing as it likely scans for PASSED but this test has multiple tests, and some of them currently fail (without this patch):

```
ctest -R hipStreamSemantics  -V
UpdateCTestConfiguration  from :/home/pjaaskel/src/chip-spv/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/pjaaskel/src/chip-spv/build/DartConfiguration.tcl
Test project /home/pjaaskel/src/chip-spv/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 1309
    Start 1309: hipStreamSemantics

1309: Test command: /home/pjaaskel/src/chip-spv/build/samples/hipStreamSemantics/hipStreamSemantics
1309: Test timeout computed to be: 10000000
1309: ------------------------------------------------------------------
1309: TestStreamSemantics_1(stream query) : PASSED
1309: callback_sleep10: Going to sleep for 10sec
1309: ------------------------------------------------------------------
1309: TestStreamSemantics_2 (non-blocking stream): PASSED
1309: ------------------------------------------------------------------
1309: Going to sync stream1
1309: Going to call query
1309: TestStreamSemantics_3(independent execution) :  Failed 
1309: ------------------------------------------------------------------
1309: Going to destroy a stream which has pending work
1309: Waiting for device to finish the task
1309: TestStreamSemantics_4 (stream destroy): PASSED
1309: ------------------------------------------------------------------
1309: TestStreamSemantics_5 (sync streams):  Failed 
1/1 Test #1309: hipStreamSemantics ...............   Passed    0.57 sec

The following tests passed:
        hipStreamSemantics

100% tests passed, 0 tests failed out of 1
```

pjaaskel commented:
> While testing, I noticed hipStreamSemantics is falsely detected as passing as it likely scans for PASSED but this test has multiple tests, and some of them currently fail (without this patch):

Added a FAIL regexp for this.

pvelesko commented:
> OK, I reproduced it successfully in your box and fixed the problem (I think).

you mentioned hipStreamSemantics but the only two tests that were failing were 

```
Unit_hipStreamAddCallback_WithDefaultStream
Unit_hipStreamAddCallback_WithCreatedStream
```

pvelesko commented:
ok seems that they're passing now

pvelesko commented:
only `cuda-convolutionSeparable` test failed for Level Zero. Merging this & will investigate that failure separately.

pjaaskel commented:
I think I didn't touch anything which should affect the L0 BE, I hope.

pjaaskel commented:
Ah, the test fix of course could cause LZ to fail if it handles it wrongly.
